### PR TITLE
Feature/room upgrade

### PR DESCRIPTION
**Pull request by ganfra**
### Pull Request Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 16
- [ ] UI change has been tested on both light and dark themes
- [ ] Pull request is based on the develop branch
- [ ] Pull request updates [CHANGES.md](https://github.com/vector-im/riotX-android/blob/develop/CHANGES.md)
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *  * Copyright 2019 New Vector Ltd
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt]: strange * * copyright?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        @Json(name = "creator") val creator: String,
+        @Json(name = "room_version") val roomVersion: String?,
+        @Json(name = "predecessor") val predecessor: Predecessor?
+)
+
+/**
+ * A link to an old room in case of room versioning
+ */
+@JsonClass(generateAdapter = true)
+data class Predecessor(
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt]: BM won't like this :)

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/tombstone/RoomTombstoneContent.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/*
+ *
+ *  * Copyright 2019 New Vector Ltd
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/tombstone/RoomTombstoneContent.kt]: again

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/tombstone/RoomTombstoneContent.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+
+/**
+ * Class to contains Tombstone information
+ */
+@JsonClass(generateAdapter = true)
+data class RoomTombstoneContent(
+        @Json(name = "body") val body: String? = null,
+        @Json(name = "replacement_room") val replacementRoom: String
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/tombstone/RoomTombstoneContent.kt]: maybe make this optional? (as per spec it's required, like body no?) but anyway we may want to resist to malformed?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/RoomSummaryEntity.kt [hunk 5]
```diff
@@ -40,12 +41,19 @@ internal open class RoomSummaryEntity(@PrimaryKey var roomId: String = "",
...
+    private var versioningStateStr: String = VersioningState.NONE.name
+
 
     @delegate:Ignore
     var membership: Membership by Delegates.observable(Membership.valueOf(membershipStr)) { _, _, newValue ->
         membershipStr = newValue.name
     }
 
+    @delegate:Ignore
+    var versioningState: VersioningState by Delegates.observable(VersioningState.valueOf(versioningStateStr)) { _, _, newValue ->
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/RoomSummaryEntity.kt]: i think i remember we had an issue using delegates for two way binding?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/RoomFactory.kt [hunk 6]
```diff
@@ -40,11 +41,14 @@ import im.vector.matrix.android.internal.session.room.timeline.DefaultTimelineSe
...
 import im.vector.matrix.android.internal.session.room.timeline.PaginationTask
 import im.vector.matrix.android.internal.task.TaskExecutor
+import io.realm.RealmConfiguration
 import javax.inject.Inject
 
 internal class RoomFactory @Inject constructor(private val context: Context,
                                                private val credentials: Credentials,
                                                private val monarchy: Monarchy,
+                                               @SessionDatabase
+                                               private val realmConfiguration: RealmConfiguration,
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/RoomFactory.kt]: it is not accessible from Monarchy?

Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/RoomFactory.kt]: can get it from Monarchy?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/state/DefaultStateService.kt [hunk 7]
```diff
@@ -17,16 +17,32 @@
...
 
 internal class DefaultStateService @Inject constructor(private val roomId: String,
+                                                       @SessionDatabase
+                                                       private val realmConfiguration: RealmConfiguration,
                                                        private val taskExecutor: TaskExecutor,
                                                        private val sendStateTask: SendStateTask) : StateService {
 
+    override fun getStateEvent(eventType: String): Event? {
+        return Realm.getInstance(realmConfiguration).use { realm ->
+            EventEntity.where(realm, roomId, eventType).prev()?.asDomain()
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/state/DefaultStateService.kt]: prev?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/tombstone/RoomTombstoneEventLiveObserver.kt [hunk 8]
```diff
@@ -0,0 +1,78 @@
+/*
+ *
+ *  * Copyright 2019 New Vector Ltd
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/tombstone/RoomTombstoneEventLiveObserver.kt]: copyright * *

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt [hunk 9]
```diff
@@ -239,12 +257,38 @@ class RoomDetailFragment :
...
 
+    private fun setupNotificationView() {
+        notificationAreaView.delegate = object : NotificationAreaView.Delegate {
+
+            override fun onTombstoneEventClicked(tombstoneEvent: Event) {
+                roomDetailViewModel.process(RoomDetailActions.HandleTombstoneEvent(tombstoneEvent))
+            }
+
+            override fun resendUnsentEvents() {
+                TODO("not implemented")
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt]: will crash? maybe snackbar?

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/TimelineEventController.kt [hunk 10]
```diff
@@ -219,8 +220,8 @@ class TimelineEventController @Inject constructor(private val dateFormatter: Tim
                 // Should be build if not cached or if cached but contains mergedHeader or formattedDay
                 // We then are sure we always have items up to date.
                 if (modelCache[position] == null
-                        || modelCache[position]?.mergedHeaderModel != null
-                        || modelCache[position]?.formattedDayModel != null) {
+                    || modelCache[position]?.mergedHeaderModel != null
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/TimelineEventController.kt]: this condition is a bit hard to understand

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/factory/RoomCreateItemFactory.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
+/*
+ *
+ *  * Copyright 2019 New Vector Ltd
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/factory/RoomCreateItemFactory.kt]: again copyright

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/RoomCreateItem.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *  * Copyright 2019 New Vector Ltd
+ *  *
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/RoomCreateItem.kt]: * *

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/RoomSummaryEntity.kt (hunk 5 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/RoomSummaryEntity.kt]: Why is there still membership so????

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/RoomFactory.kt (hunk 6 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/RoomFactory.kt]: Yep, will fix

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt (hunk 9 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt]: Won't as it's not used, imported from Riot Legacy

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/TimelineEventController.kt (hunk 10 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/TimelineEventController.kt]: it hasn't change in this PR

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt (hunk 2 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/create/RoomCreateContent.kt]: Why? having 2 classes in one file? :P

**Pull request merged**